README.md
# cz-git-universal-jira-conventional

**cz-git-universal-jira-conventional** is a plugin for the [**commitizen tools**](https://github.com/commitizen-tools/commitizen), a toolset that helps you to create [**conventional commit messages**](https://www.conventionalcommits.org/en/v1.0.0/). Since the structure of conventional commits messages is standardized they are machine readable and allow commitizen to automaticially calculate and tag [**semantic version numbers**](https://semver.org/) as well as create **CHANGELOG.md** files for your releases.

This plugin extends the commitizen tools by:
- **adding a Jira issue id** in the commit message (can be optional, see config below)
- **create links to Git-Repository** commits in the CHANGELOG.md
- **create links to Jira** issues in the CHANGELOG.md

This plugin is a fork of the original [cz-github-jira-conventional plugin](https://github.com/apheris/cz-github-jira-conventional).
Without it, this plugin would not be possible. It extends it by adding generic git commit path support for different git platforms and the option for leaving the jira issue id empty.

When you call commitizen `commit` you will be prompted to enter the scope of your commit as a Jira issue id (or multiple issue ids, prefixed or without prefix, see config below).
```
> cz commit
? Select the type of change you are committing fix: A bug fix. Correlates with PATCH in SemVer
? JIRA issue number (multiple "42, 123"). XX-
...
```

The changelog created by cz (`cz bump --changelog`)will contain links to the commits in Github and the Jira issues.
```markdown
## v1.0.0 (2021-08-06)

### Features

- **[XX-123](https://myproject.atlassian.net/browse/XX-123)**: create changelogs with links to issues and commits [a374b](https://github.com/apheris/cz-github-jira-conventional/commit/a374b93f39327964f5ab5290252b795647906008)
- **[XX-42](https://myproject.atlassian.net/browse/XX-42),[XX-13](https://myproject.atlassian.net/browse/XX-13)**: allow multiple issue to be referenced in the commit [07ab0](https://github.com/apheris/cz-github-jira-conventional/commit/07ab0e09de36712ab1db93fff0c821ecd80b5849)
``` 


## Installation

Install with pip
`python -m pip install cz-git-universal-jira-conventional` 

You need to use a cz config file that has the **required** additional values `jira_base_url` and `git_commit_base_url` and may contain the **optional** value `jira_prefix`.

Example `.cz.yaml` config for this repository
```yaml
commitizen:
  name: cz_git_universal_jira_conventional
  tag_format: v$version
  version: 1.0.0
  jira_prefix: XX-
  jira_base_url: https://myproject.atlassian.net
  git_commit_base_url: https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/
```

The `jira_prefix` can be either 
- empty (the user must write the prefix for each issue)
- a string (the prefix will be added automatically)
- a list (for multiple projects, the user will be asked to choose a prefix)

```yaml
  jira_prefix: 
    - XX-
    - XY-
    - YY-
```

The `jira_allow_empty` can be either
- `false` (default, a jira issue id is mandatory)
- `true` (jira issue id is optional and can be omitted)

The `git_commit_base_url` depends on your git platform vendor.
You want to add an url which can be appended with to commit-sha for reference. Examples:
- Github: `https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/`
- Bitbucket (Self-Hosted): `https://bitbucket.example.com/projects/PROJECT-NAME/repos/REPO-NAME/commits/`

### pre-commit
Add this plugin to the dependencies of your commit message linting with `pre-commit`. 

Example `.pre-commit-config.yaml` file.
```yaml
repos:
  - repo: https://github.com/commitizen-tools/commitizen
    rev: v2.17.13
    hooks:
      - id: commitizen
        stages: [commit-msg]
        additional_dependencies: [cz-git-universal-jira-conventional]
```
Install the hook with 
```bash
pre-commit install --hook-type commit-msg
```

<!-- LICENSE -->
## License

Distributed under the MIT License. See [LICENSE](LICENSE) for more information.

<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements
This plugin is a fork of the original [cz-github-jira-conventional plugin](https://github.com/apheris/cz-github-jira-conventional).
Without it, this plugin would not be possible.

This plugin would not have been possible without the fantastic work from:
* [commitizen tools](https://github.com/commitizen-tools/commitizen)
* [conventional_JIRA](https://github.com/Crystalix007/conventional_jira)
* [cz-github-jira-conventional plugin](https://github.com/apheris/cz-github-jira-conventional)

CHANGELOG.md
## v1.2.0 (2022-10-26)

### Feat

- refactored project [3957b](https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/3957bd4ecc79f0ff50e591ddc9812de478eb1899)
- added universal git project support [5bbff](https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/5bbffd775b831a531179a58cd474a0c3215e230d)

## v1.1.0 (2022-07-26)

### Feat

- **[XX-2815](https://myproject.atlassian.net/browse/XX-2815)**: allow choosing between multiple Jira projects [469b9](https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/469b94c3bb3aa61c6b8c53627c064e5921b4d912)

## test (2021-08-16)

### Feat

- **[XX-443](https://myproject.atlassian.net/browse/XX-443)**: create cz customization script that links github and jira [271e7](https://github.com/ngoldack/cz-git-universal-jira-conventional/commit/271e78a3d8505192615702434ef9839b2ef3c08c)


cz_git_universal_jira_conventional.py
import os
import re
from typing import Any, Dict, List

from commitizen import defaults, git, config
from commitizen.cz.base import BaseCommitizen
from commitizen.cz.utils import multiple_line_breaker, required_validator
from commitizen.cz.exceptions import CzException

__all__ = ["GitJiraConventionalCz"]


def parse_subject(text):
    if isinstance(text, str):
        text = text.strip(".").strip()

    return required_validator(text, msg="Subject is required.")


class GitJiraConventionalCz(BaseCommitizen):
    bump_pattern = defaults.bump_pattern
    bump_map = defaults.bump_map
    commit_parser = defaults.commit_parser
    changelog_pattern = defaults.bump_pattern

    # Read the config file and check if required settings are available
    conf = config.read_cfg()
    jira_prefix_hint = ""
    if "jira_prefix" in conf.settings:
        jira_prefix = conf.settings["jira_prefix"]
        issue_multiple_hint = "42, 123"
        # if there is only one project prefix, show it as prefix hint
        if not isinstance(jira_prefix, list):
            jira_prefix_hint = jira_prefix
    else:
        jira_prefix = ""
        issue_multiple_hint = "XZ-42, XY-123"

    if "jira_base_url" not in conf.settings:
        print(
            "Please add the key jira_base_url to your .cz.yaml|json|toml config file."
        )
        quit()
    if "git_commit_base_url" not in conf.settings:
        print("Please add the key git_commit_base_url to your .cz.yaml|json|toml config file.")
        quit()
    jira_base_url = conf.settings["jira_base_url"]
    git_commit_base_url = conf.settings["git_commit_base_url"]
    jira_allow_empty = conf.settings["jira_allow_empty"]

    # if "change_type_map" not in conf.settings:
    change_type_map = {
        "feat": "Feat",
        "fix": "Fix",
        "refactor": "Refactor",
        "perf": "Perf",
    }

    def questions(self) -> List[Dict[str, Any]]:
        questions: List[Dict[str, Any]] = [
            {
                "type": "list",
                "name": "prefix",
                "message": "Select the type of change you are committing",
                "choices": [
                    {
                        "value": "fix",
                        "name": "fix: A bug fix. Correlates with PATCH in SemVer",
                    },
                    {
                        "value": "feat",
                        "name": "feat: A new feature. Correlates with MINOR in SemVer",
                    },
                    {"value": "docs", "name": "docs: Documentation only changes"},
                    {
                        "value": "style",
                        "name": (
                            "style: Changes that do not affect the "
                            "meaning of the code (white-space, formatting,"
                            " missing semi-colons, etc)"
                        ),
                    },
                    {
                        "value": "refactor",
                        "name": (
                            "refactor: A code change that neither fixes "
                            "a bug nor adds a feature"
                        ),
                    },
                    {
                        "value": "perf",
                        "name": "perf: A code change that improves performance",
                    },
                    {
                        "value": "test",
                        "name": (
                            "test: Adding missing or correcting " "existing tests"
                        ),
                    },
                    {
                        "value": "build",
                        "name": (
                            "build: Changes that affect the build system or "
                            "external dependencies (example scopes: pip, docker, npm)"
                        ),
                    },
                    {
                        "value": "ci",
                        "name": (
                            "ci: Changes to our CI configuration files and "
                            "scripts (example scopes: GitLabCI)"
                        ),
                    },
                ],
            },
            {
                "type": "input",
                "name": "scope",
                "message": (
                    f'JIRA issue number (multiple "{self.issue_multiple_hint}"). {self.jira_prefix_hint}'
                ),
                "filter": self.parse_scope,
            },
            {
                "type": "input",
                "name": "subject",
                "filter": parse_subject,
                "message": (
                    "Write a short and imperative summary of the code changes: (lower case and no period)\n"
                ),
            },
            {
                "type": "input",
                "name": "body",
                "message": (
                    "Provide additional contextual information about the code changes: (press [enter] to skip)\n"
                ),
                "filter": multiple_line_breaker,
            },
            {
                "type": "confirm",
                "message": "Is this a BREAKING CHANGE? Correlates with MAJOR in SemVer",
                "name": "is_breaking_change",
                "default": False,
            },
            {
                "type": "input",
                "name": "footer",
                "message": (
                    "Footer. Information about Breaking Changes and "
                    "reference issues that this commit closes: (press [enter] to skip)\n"
                ),
            },
        ]
        # If there are multiple Jira prefixes let the user select one
        if isinstance(self.jira_prefix, list):
            questions.insert(
                1,
                {
                    "type": "list",
                    "name": "issue_jira_prefix",
                    "message": "JIRA project",
                    "choices": [
                        {"value": prefix, "name": prefix} for prefix in self.jira_prefix
                    ],
                },
            )
        return questions

    def parse_scope(self, text):
        if self.jira_allow_empty and len(text) == 0:
            return ""

        if self.jira_prefix:
            issueRE = re.compile(r"\d+")
        else:
            issueRE = re.compile(r"\w+-\d+")

        if not text:
            return ""

        issues = [i.strip() for i in text.strip().split(",")]
        for issue in issues:
            if not issueRE.fullmatch(issue):
                raise InvalidAnswerError(f"JIRA issue '{issue}' is not valid.")

        return required_validator(issues, msg="JIRA scope is required")

    def message(self, answers: dict) -> str:
        prefix = answers["prefix"]
        issue_jira_prefix = (
            answers["issue_jira_prefix"]
            if "issue_jira_prefix" in answers
            else self.jira_prefix
        )
        issues = answers["scope"]
        subject = answers["subject"]
        body = answers["body"]
        footer = answers["footer"]
        is_breaking_change = answers["is_breaking_change"]
        scope = ""

        if issues:
            # Add Jira prefixes to the issue numbers.
            issues_str = ",".join([issue_jira_prefix + i for i in issues])
            scope = f"({issues_str})"
        if body:
            body = f"\n\n{body}"
        if is_breaking_change:
            footer = f"BREAKING CHANGE: {footer}"
        if footer:
            footer = f"\n\n{footer}"

        message = f"{prefix}{scope}: {subject}{body}{footer}"

        return message

    def example(self) -> str:
        return (
            "fix: correct minor typos in code\n"
            "\n"
            "see the issue for details on the typos fixed\n"
            "\n"
            "closes issue #12"
        )

    def schema(self) -> str:
        return (
            "<type>(<scope>): <subject>\n"
            "<BLANK LINE>\n"
            "<body>\n"
            "<BLANK LINE>\n"
            "(BREAKING CHANGE: )<footer>"
        )

    def schema_pattern(self) -> str:
        PATTERN = (
            r"(build|ci|docs|feat|fix|perf|refactor|style|test|chore|revert|bump)"
            r"(\(\S+\))?!?:(\s.*)"
        )
        return PATTERN

    def info(self) -> str:
        dir_path = os.path.dirname(os.path.realpath(__file__))
        filepath = os.path.join(dir_path, "conventional_commits_info.txt")
        with open(filepath, "r") as f:
            content = f.read()
        return content

    def process_commit(self, commit: str) -> str:
        pat = re.compile(self.schema_pattern())
        m = re.match(pat, commit)
        if m is None:
            return ""
        return m.group(3).strip()

    def changelog_message_builder_hook(
        self, parsed_message: dict, commit: git.GitCommit
    ) -> dict:
        """add git and jira links to the readme"""
        rev = commit.rev
        m = parsed_message["message"]
        if parsed_message["scope"]:
            parsed_message["scope"] = " ".join(
                [
                    f"[{issue_id}]({self.jira_base_url}/browse/{issue_id})"
                    for issue_id in parsed_message["scope"].split(",")
                ]
            )

        # Remove trailing slash
        if self.git_commit_base_url[-1] == "/":
            self.git_commit_base_url = self.git_commit_base_url[:-1]
        parsed_message[
            "message"
        ] = f"{m} [{rev[:5]}]({self.git_commit_base_url}/{commit.rev})"
        return parsed_message


class InvalidAnswerError(CzException):
    ...


discover_this = GitJiraConventionalCz
